### DISCUSSION TITLE
How to get and display the analytic event count from app

### DISCUSSION THREAD

**Discussion by anhnch** [invertase/react-native-firebase#6428] [Q&A] [answered]
I'm using @react-native-firebase/analytics in my app. I can log and the number a document is displayed. But is there anyway to get the count from the app and display it as view number?

Of course I can manage the view number on my own by updating the document in firestore every time it is viewed. But it'll be more convenient to get it from analytics instead. This reduces the number of write and locking the document when writing, especially when there are too many active users accessing document. Updating it too frequently just for view number is not a good idea.

But it seems like analytics is a "one way" write-only package.

**Comment by mikehardy** [comment, marked as answer]
indeed - analytics is a one-way write-only package. There is no ability at all to count. You can wrap it if you want and use some local state management solution for a specific local device, but that's it. Or do something on the backend that uses analytics export to bigquery and checks through bigquery, in perhaps a firebase function. I'm not sure what your use case is though, so that may not fit at all.

**Reply by anhnch** [reply]
My use case is very straightforward. My app has articles, when a user clicks to read an article, its view count should be increased by 1. That's it.

And I'm looking for a best practice to manage this view count. Constantly updating the article to increase the count is a bad idea.

At least it's good to have your confirmation that analytics is write-only package.

I think I'll have to store the count in a separated documents for stats instead, to avoid too much writing to the articles. I don't want mess with Bigquery and function just for a count. Or I can drop the feature.

I guess I have to open a question in Stackoverflow, because it's not react-native-firebase related anymore.

Thank you!

**Reply by mikehardy** [reply]
There is a standard solution for this in the docs, though as with all things there are tradeoffs: https://cloud.google.com/firestore/docs/solutions/counters

**Reply by anhnch** [reply]
I've never thought that they discuss this among 13 solutions in the official docs :) Anyway the standard solution is a bit overkill to my small scale app, which is not count sensitive. 

I'm thinking of implementing an API to increase the count in Redis and update the firestore document every 10 or 15 minutes. This solves 3 main things:

- Writing to Redis doesn't lock firestore article document. Redis is fast.
- Cron updating the article document count field every 10 minutes is acceptable to my app.
- When I get a list of articles I have the count field along the way.

For the scalable standard solution:

- It's a bit cumbersome sharding with performance and cost tradeoffs
- We have to clean up the shards when article is deleted.
- It's more suitable for comments/likes count, where we need to know who and when users liked/commented. While for view count (we don't need to know who read), the sub-collection documents contains no information.
- We still have to sync the shard number to the article.
- Reading an article with view count requires transaction to count and sum the shards.
- You don't have view count when we get list of articles

(Sorry for getting you into the problem and thank you for your time)
